File: flask/src/validation/user_validation.py
from pydantic import BaseModel, EmailStr, Field, field_validator
from datetime import date

class CreateUserValidation(BaseModel):
    nama: str = Field(..., max_length=100, description="Nama harus berupa teks")
    email: EmailStr = Field(..., description="Email tidak valid")
    password: str = Field(..., min_length=6, description="Password minimal harus memiliki 6 karakter")
    NIK: str = Field(..., pattern=r"^\d{16}$", description="NIK harus terdiri dari 16 karakter dan hanya angka")
    alamat: str = Field(..., max_length=100, description="Alamat tidak boleh lebih dari 100 karakter")
    telepon: str = Field(..., pattern=r"^\d{1,15}$", description="Nomor telepon harus terdiri dari angka maksimal 15 karakter")
    jenis_kelamin: str = Field(..., pattern=r"^(L|P)$", description="Jenis kelamin harus 'L' (Laki-laki) atau 'P' (Perempuan)")
    kepala_keluarga: int = Field(..., description="Kepala keluarga harus bernilai 0 atau 1")
    tempat_lahir: str = Field(..., max_length=50, description="Tempat lahir tidak boleh lebih dari 50 karakter")
    tanggal_lahir: date = Field(..., description="Tanggal lahir harus dalam format yang valid (YYYY-MM-DD)")
    jenis_usaha: str = Field(..., max_length=50, description="Jenis usaha tidak boleh lebih dari 50 karakter")

    @field_validator('kepala_keluarga')
    def validate_kepala_keluarga(cls, value):
        if value not in [0, 1]:
            raise ValueError('Kepala keluarga harus bernilai 0 atau 1')
        return value

class LoginUserValidation(BaseModel):
    email: EmailStr = Field(..., description="Email tidak valid")
    password: str = Field(..., min_length=6, description="Password minimal harus memiliki 6 karakter")

class UpdateUserValidation(BaseModel):
    nama: str = Field(None, max_length=100, description="Nama harus berupa teks dan maksimal 100 karakter")
    email: EmailStr = Field(None, description="Email tidak valid")
    password: str = Field(None, min_length=6, description="Password minimal harus memiliki 6 karakter")
    NIK: str = Field(None, pattern=r"^\d{16}$", description="NIK harus terdiri dari 16 karakter dan hanya angka")
    alamat: str = Field(None, max_length=100, description="Alamat tidak boleh lebih dari 100 karakter")
    telepon: str = Field(None, pattern=r"^\d{1,15}$", description="Nomor telepon harus terdiri dari angka maksimal 15 karakter")
    jenis_kelamin: str = Field(None, pattern=r"^(L|P)$", description="Jenis kelamin harus 'L' (Laki-laki) atau 'P' (Perempuan)")
    kepala_keluarga: int = Field(None, description="Kepala keluarga harus bernilai 0 atau 1")
    tempat_lahir: str = Field(None, max_length=50, description="Tempat lahir tidak boleh lebih dari 50 karakter")
    tanggal_lahir: date = Field(None, description="Tanggal lahir harus dalam format yang valid (YYYY-MM-DD)")
    jenis_usaha: str = Field(None, max_length=50, description="Jenis usaha tidak boleh lebih dari 50 karakter")

    @field_validator('kepala_keluarga')
    def validate_kepala_keluarga(cls, value):
        if value is not None and value not in [0, 1]:
            raise ValueError('Kepala keluarga harus bernilai 0 atau 1')
        return value

File: express/src/validation/user-validation.js
import Joi from "joi";

const createUserValidation = Joi.object({
  nama: Joi.string().max(100).required().messages({
    "string.base": "Nama harus berupa teks",
    "string.max": "Nama tidak boleh lebih dari 100 karakter",
    "any.required": "Nama harus diisi",
  }),
  email: Joi.string().email({ minDomainSegments: 2 }).required().messages({
    "string.email": "Email tidak valid",
    "any.required": "Email harus diisi",
  }),
  password: Joi.string().min(6).required().messages({
    "string.min": "Password minimal harus memiliki 6 karakter",
    "any.required": "Password harus diisi",
  }),
  NIK: Joi.string().length(16).regex(/^\d+$/).required().messages({
    "string.length": "NIK harus terdiri dari 16 karakter",
    "string.pattern.base": "NIK hanya boleh terdiri dari angka",
    "any.required": "NIK harus diisi",
  }),
  alamat: Joi.string().max(100).required().messages({
    "string.max": "Alamat tidak boleh lebih dari 100 karakter",
    "any.required": "Alamat harus diisi",
  }),
  telepon: Joi.string().max(15).regex(/^\d+$/).required().messages({
    "string.max": "Nomor telepon tidak boleh lebih dari 15 karakter",
    "string.pattern.base": "Nomor telepon hanya boleh terdiri dari angka",
    "any.required": "Nomor telepon harus diisi",
  }),
  jenis_kelamin: Joi.string().valid("L", "P").required().messages({
    "any.only": "Jenis kelamin harus 'L' (Laki-laki) atau 'P' (Perempuan)",
    "any.required": "Jenis kelamin harus diisi",
  }),
  kepala_keluarga: Joi.number().integer().valid(0, 1).required().messages({
    "any.only": "Kepala keluarga harus bernilai 0 atau 1",
    "number.base": "Kepala keluarga harus berupa angka",
    "any.required": "Field kepala keluarga wajib diisi",
  }),
  tempat_lahir: Joi.string().max(50).required().messages({
    "string.max": "Tempat lahir tidak boleh lebih dari 50 karakter",
    "any.required": "Tempat lahir harus diisi",
  }),
  tanggal_lahir: Joi.date().iso().required().messages({
    "date.format": "Tanggal lahir harus dalam format yang valid (YYYY-MM-DD)",
    "any.required": "Tanggal lahir harus diisi",
  }),
  jenis_usaha: Joi.string().max(50).required().messages({
    "string.max": "Jenis usaha tidak boleh lebih dari 50 karakter",
    "any.required": "Jenis usaha harus diisi",
  }),
});

const loginUserValidation = Joi.object({
  email: Joi.string().email({ minDomainSegments: 2 }).required().messages({
    "string.email": "Email tidak valid",
    "any.required": "Email harus diisi",
  }),
  password: Joi.string().min(6).required().messages({
    "string.min": "Password minimal harus memiliki 6 karakter",
    "any.required": "Password harus diisi",
  }),
});

const updateUserValidation = Joi.object({
  nama: Joi.string().max(100).optional().messages({
    "string.base": "Nama harus berupa teks",
    "string.max": "Nama tidak boleh lebih dari 100 karakter",
  }),
  email: Joi.string().email({ minDomainSegments: 2 }).optional().messages({
    "string.email": "Email tidak valid",
  }),
  password: Joi.string().min(6).optional().messages({
    "string.min": "Password minimal harus memiliki 6 karakter",
  }),
  NIK: Joi.string().length(16).regex(/^\d+$/).optional().messages({
    "string.length": "NIK harus terdiri dari 16 karakter",
    "string.pattern.base": "NIK hanya boleh terdiri dari angka",
  }),
  alamat: Joi.string().max(100).optional().messages({
    "string.max": "Alamat tidak boleh lebih dari 100 karakter",
  }),
  telepon: Joi.string().max(15).regex(/^\d+$/).optional().messages({
    "string.max": "Nomor telepon tidak boleh lebih dari 15 karakter",
    "string.pattern.base": "Nomor telepon hanya boleh terdiri dari angka",
  }),
  jenis_kelamin: Joi.string().valid("L", "P").optional().messages({
    "any.only": "Jenis kelamin harus 'L' (Laki-laki) atau 'P' (Perempuan)",
  }),
  kepala_keluarga: Joi.number().integer().valid(0, 1).optional().messages({
    "any.only": "Kepala keluarga harus bernilai 0 atau 1",
    "number.base": "Kepala keluarga harus berupa angka",
  }),
  tempat_lahir: Joi.string().max(50).optional().messages({
    "string.max": "Tempat lahir tidak boleh lebih dari 50 karakter",
  }),
  tanggal_lahir: Joi.date().iso().optional().messages({
    "date.format": "Tanggal lahir harus dalam format yang valid (YYYY-MM-DD)",
  }),
  jenis_usaha: Joi.string().max(50).optional().messages({
    "string.max": "Jenis usaha tidak boleh lebih dari 50 karakter",
  }),
});

export { createUserValidation, loginUserValidation, updateUserValidation };
